### ISSUE TITLE
50×55 动作输出的关节索引映射与补零对齐策略

### ISSUE THREAD

**Issue by lyhgalaxy** [Robbyant/lingbot-vla-v2#35]
Hi，非常感谢您开源 LingBot‑VLA‑v2 项目！在复现使用预训练权重过程中遇到两个疑问，想向您请教下：

1. 模型输出动作维度是 `50×55`，我想确认下这**55 维动作向量内部各个位置和机器人本体关节、夹爪 / 末端执行器的对应映射关系**。比如哪些索引对应机械臂各关节、哪些对应手部 / 夹爪，类似 “索引 0‑7 对应手臂关节” 这种分段说明。
2. 训练数据集里不同机器人手部自由度并不统一：有的机器人手部是 6 自由度，有的是 7 自由度。想了解预训练阶段是怎么对齐不同自由度的？是**按照部位维度原地补 0**，还是统一在动作向量的**末尾位置追加补 0**来做维度对齐？

**Comment by LuFan31** [comment]
Thanks for your interest in LingBot-VLA 2.0! There are two levels of padding/mapping involved here.

### 1. Mapping of the `50 × 55` action output

Here, `50` is the action chunk length, while `55` is the model's `max_action_dim`.

The 55-dimensional action vector is not hard-coded in the model as a fixed joint-index table. Instead, the **order and maximum dimension of each body-part feature are defined by `data.joints` in the training configuration**.

For example, the current [real_robot.yaml](https://github.com/Robbyant/lingbot-vla-v2/blob/main/configs/vla/real_robot/real_robot.yaml) defines:

```yaml
joints:
  - arm.position: 14
  - end.position: 14
  - effector.position: 2
  - waist.position: 4
  - head.position: 2
  - base.position: 3
  - hand.position: 12
```

Therefore, for this configuration, the concatenated layout is:

```text
[ 0:14)  arm.position
[14:28)  end.position
[28:30)  effector.position
[30:34)  waist.position
[34:36)  head.position
[36:39)  base.position
[39:51)  hand.position
[51:55)  padding / reserved dimensions
```

The mapping **within each body-part segment** to the raw robot joints is dataset/embodiment-specific and is defined by the corresponding `robot_config`, through its `origin_keys` and `start` / `end` indices.

Therefore, when adapting the model to your own robot, please use the `data.joints` configuration together with the corresponding `robot_config`.

### 2. Alignment of embodiments with different DoFs

The implementation performs **body-part-wise padding**, rather than concatenating all valid DoFs first and only padding at the end.

For each joint type, `pad_and_concat()` pads the feature to the maximum dimension specified in `data.joints`. For example, with:

```yaml
- arm.position: 14
```

if an embodiment has a 7-dimensional `arm.position`, its **arm segment** is represented as:

```text
[7 valid dimensions | 7 padded dimensions]
```

while a 12-dimensional arm would be:

```text
[12 valid dimensions | 2 padded dimensions]
```

The corresponding `action_joint_mask` marks the real dimensions as valid and the padded dimensions as invalid. These padded dimensions are excluded from the action loss during training.

During inference, the same joint mask is used to remove padded dimensions before the action is mapped back to the original robot-specific representation.

**Comment by lyhgalaxy** [comment]
Thank you for the detailed explanation! I have one more question regarding the dexterous-hand representation.

I noticed that the unified action space defines:

```yaml
- hand.position: 12
```

and the README describes these 12 dimensions as **hand joint positions**.

However, I could not find a public `robot_config` example that contains `hand.position`, so I would like to clarify what these 12 dimensions correspond to in the pre-training data.

1. Does `hand.position` represent the **physical finger joint angles**, or the **motor / actuator positions** of the dexterous hand?

2. If the dexterous hand is tendon-driven or contains coupled joints, is the motor-to-joint mapping handled before the data is converted into `hand.position`, or does the model directly predict actuator positions?

3. For the 12-dimensional canonical `hand.position`, how many hands and fingers does it represent? For example, is it:

   * one dexterous hand with up to 12 DoFs,
   * two hands with up to 6 DoFs each,
   * or simply a maximum per-body-part dimension shared across different embodiments?

4. Could you provide an example from one of the dexterous-hand embodiments used during pre-training, showing the mapping, for example:

```text
hand.position[0:x]  -> left/right hand, thumb
hand.position[x:y]  -> index finger
...
```

or alternatively the corresponding `robot_config`?

This would be very helpful for understanding how to map the pretrained model's `hand.position` output to the low-level controller of a custom dexterous hand.

Thanks again!

**Comment by myshrry** [comment]
Also, did LingBotVLA-v2 pre-training include any end-effector pose data in axis-angle form, or is it strictly XYZ + quaternion (7d per arm)? I'm  post-training on LIBERO whose raw actions are axis-angle, and want to know if converting to quaternion is required to get better result, or if I can feed axis-angle directly.
